gaps/gap-draft_experimental_features/gap-draft_experimental_features.md

```diff
@@ -0,0 +1,263 @@
+
+(TODO: Check if we can use `!` sign. May cause conflict in constraint language,
+json or yaml Offer/Demand representation)
+
+We have multiple options of correct experimental namespaces:
+- `"golem.com.payment.!exp.payment-platform"`
+- `"golem.com.!exp.payment.payment-platform"`
+- `"golem.!exp.com.payment.payment-platform"`
+- `"!exp.golem.com.payment.payment-platform"`
+
+Example of incorrect namespacing:
+
+`"golem.com.payment.payment-platform.!exp" : "erc20-rinkeby-tglm"`
+
+`payment-platform` property didn't exist before, so we are not allowed to introduce it.
+
+#### GSB protocols
+
+GSB allows to make RPC calls to other yagna daemons. Messages are sent to specific address
+like for example:
+`/public/market/protocol/mk1/discovery/offers/GetOffers`
+
+Examples of introducing new GSB message `ListOffers`:
+- `/public/!exp/market/protocol/mk1/discovery/offers/ListOffers`
+- `/public/market/!exp/protocol/mk1/discovery/offers/ListOffers`
+- `/public/market/protocol/!exp/mk1/discovery/offers/ListOffers`
+- `/public/market/protocol/mk1/!exp/discovery/offers/ListOffers`
+- `/public/market/protocol/mk1/discovery/!exp/offers/ListOffers`
+- `/public/market/protocol/mk1/discovery/offers/!exp/ListOffers`
+
+Note that `/public` prefix has special meaning. Net module redirects all calls to `/public`
+addresses. For this reason `/!exp/public` is not correct namespacing.
+
+##### Backoff to stable version
+
+In case of GSB messages it is hard to imagine, that changes will be completely independent
+of previous versions of protocols. In this case it might be useful to check if other yagna
+daemon supports experimental version and backoff to stable code if it doesn't.
+
+We would recommend always binding `IsSupported` endpoint, when developing new **Experimental Feature**.
+In combination with [versioning](#Versioning-scheme) it could allow us to send GSB message
+to check if other yagna supports this specific experimental feature.
+Sending message will result with `Endpoint not found` in case it is not supported.
+
+#### REST APIs
+
+Example address for responding to Proposal:
+`POST /market-api/v1/demands/{demand_id}}/proposals/{proposal_id}`
+
+Examples of experimental namespacing:
+- `POST /market-api/v1/!exp/demands/{demand_id}}/proposals/{proposal_id}`
+- `POST /market-api/v1/demands/!exp/{demand_id}}/proposals/{proposal_id}`
+- `POST /market-api/v1/demands/{demand_id}}/!exp/proposals/{proposal_id}`
+- `POST /market-api/v1/demands/{demand_id}}/proposals/!exp/{proposal_id}`
+
+#### CLIs and other user facing interfaces
+
+- Commands should be explicitly classified as **Stable** or **Experimental**
+- We should introduce new commands instead of changing previous, when introducing
+**Experimental Feature**
+
+In general, I would treat all CLI commands that we have now, as experimental and slowly
+move them to stable. I don't think CLI stability is important for us at this point.
+
+#### SDKs
+
+SDK features are outside the scope of this GAP.
```

Comment:
Actually I feel that including guidelines for SDK libraries should be fairly straightforward, here's a draft proposal:

Experimental features shall be clearly indicated in SDK APIs in a similar way to Deprecated features, ie:
- mark experimental elements of APIs using a method standard for the language of the SDK library (eg. `@experimental` decorator in Python)
- once a feature is promoted to stable status, remove the 'experimental' mark from the respective APIs.